src/server/wesnothd/server.cpp

```diff
@@ -880,17 +880,23 @@ template<class SocketPtr> bool server::authenticate(
 {
 	// Current login procedure  for registered nicks is:
 	// - Client asks to log in with a particular nick
-	// - Server sends client random nonce plus some info
-	// 	generated from the original hash that is required to
-	// 	regenerate the hash
-	// - Client generates hash for the user provided password
-	// 	and mixes it with the received random nonce
-	// - Server received password hash hashed with the nonce,
-	// applies the nonce to the valid hash and compares the results
+	// - Server sends client a password request (if TLS/database support is enabled)
+	// - Client sends the plaintext password
+	// - Server receives plaintext password, hashes it, and compares it to the password in the forum database
 
 	registered = false;
 
 	if(user_handler_) {
+		const std::string salt = user_handler_->extract_salt(username);
+		if(salt.length() < 2) {
```

Comment:
Where does the 2 come from?

Reply:
Later it checks whether `salt[1] == '2'`, which is UB if the salt is length 0 or length 1.  This check is also done here just so I don't need to make the hash_password() method templated just to send the error message back.

Reply:
`hash_password` has a lot of error conditions anyway. Wouldn't it just be one more case where an empty pair is returned? Or is it clear that if the salt is a single character the "login to forum" hint is more appropriate?

Reply:
I don't know which is more appropriate, I'm just preserving the existing functionality of sending that error message back to the user.  If that's not needed anymore, then I can delete that error message and instead let it fall back to the default "Password hashing failed" error message.

Reply:
But existing functionality is to send this error message only if `salt` is empty. Now it is also sent if `salt` contains 1 character.

Reply:
So I should change it check for salt being empty?

Reply:
I think so. That would get rid of the magic number 2 that can only be explained by code in another function which does its own check anyway.

Reply:
Updated.